service/clients/oracle/daemon.go

```diff
@@ -0,0 +1,164 @@
+package oracle
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"cosmossdk.io/log"
+	"github.com/skip-mev/slinky/oracle/config"
+	"github.com/skip-mev/slinky/service/servers/oracle/types"
+	"google.golang.org/grpc"
+)
+
+var _ OracleClient = (*PriceDaemon)(nil)
+
+type PriceDaemon struct {
+	logger log.Logger
+
+	// config is the configuration of the daemon.
+	config config.AppConfig
+	// client is the underlying oracle client used to fetch prices.
+	OracleClient
+	// latestResponse is the latest price response fetched by the daemon.
+	resp ThreadSafeResponse
+	// doneCh is a channel that is closed when the daemon is stopped.
+	doneCh chan struct{}
+}
+
+// NewPriceDaemon creates a new price daemon with the given configuration.
+func NewPriceDaemon(
+	logger log.Logger,
+	cfg config.AppConfig,
+	client OracleClient,
+) (*PriceDaemon, error) {
+	if logger == nil {
+		return nil, fmt.Errorf("logger cannot be nil")
+	}
+
+	if err := cfg.ValidateBasic(); err != nil {
+		return nil, err
+	}
+
+	if client == nil {
+		return nil, fmt.Errorf("oracle client cannot be nil")
+	}
+
+	return &PriceDaemon{
+		logger:       logger.With("process", "price_daemon"),
+		config:       cfg,
+		OracleClient: client,
+		doneCh:       make(chan struct{}),
+	}, nil
+}
+
+// Start starts the price daemon. This method will block until the daemon is stopped.
+func (d *PriceDaemon) Start(ctx context.Context) error {
+	if err := d.OracleClient.Start(ctx); err != nil {
+		return err
+	}
+	defer d.OracleClient.Stop()
+
+	ticker := time.NewTicker(d.config.Interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			d.logger.Info("stopping price daemon from context")
+			return ctx.Err()
+		case <-d.doneCh:
+			d.logger.Info("price daemon stopped")
+			close(d.doneCh)
```

Comment:
I would close the channel in the `Stop()` function.  This sends a final signal on the channel so this case will still be met